executor/aggregate.go

```diff
@@ -39,24 +40,107 @@ type HashAggExec struct {
 	aggCtxsMap    aggCtxsMapper
 	groupMap      *mvmap.MVMap
 	groupIterator *mvmap.Iterator
+	inputRow      chunk.Row
+	mutableRow    chunk.MutRow
+	rowBuffer     []types.Datum
 	GroupByItems  []expression.Expression
 }
 
 // Close implements the Executor Close interface.
 func (e *HashAggExec) Close() error {
+	if err := e.baseExecutor.Close(); err != nil {
+		return errors.Trace(err)
+	}
 	e.groupMap = nil
 	e.groupIterator = nil
 	e.aggCtxsMap = nil
-	return errors.Trace(e.children[0].Close())
+	return nil
 }
 
 // Open implements the Executor Open interface.
 func (e *HashAggExec) Open(goCtx goctx.Context) error {
+	if err := e.baseExecutor.Open(goCtx); err != nil {
+		return errors.Trace(err)
+	}
 	e.executed = false
 	e.groupMap = mvmap.NewMVMap()
 	e.groupIterator = e.groupMap.NewIterator()
 	e.aggCtxsMap = make(aggCtxsMapper, 0)
-	return errors.Trace(e.children[0].Open(goCtx))
+	e.inputRow = e.childrenResults[0].End()
+	e.mutableRow = chunk.MutRowFromTypes(e.Schema().GetTypes())
+	e.rowBuffer = make([]types.Datum, 0, e.Schema().Len())
+	return nil
+}
+
+// NextChunk implements the Executor NextChunk interface.
+func (e *HashAggExec) NextChunk(goCtx goctx.Context, chk *chunk.Chunk) error {
+	// In this stage we consider all data from src as a single group.
+	if !e.executed {
+		err := e.execute(goCtx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if (e.groupMap.Len() == 0) && len(e.GroupByItems) == 0 {
+			// If no groupby and no data, we should add an empty group.
+			// For example:
+			// "select count(c) from t;" should return one row [0]
+			// "select count(c) from t group by c1;" should return empty result set.
+			e.groupMap.Put([]byte{}, []byte{})
+		}
+		e.executed = true
+	}
+	chk.Reset()
+	for {
+		groupKey, _ := e.groupIterator.Next()
+		if groupKey == nil {
+			return nil
+		}
+		aggCtxs := e.getContexts(groupKey)
+		e.rowBuffer = e.rowBuffer[:0]
+		for i, af := range e.AggFuncs {
+			e.rowBuffer = append(e.rowBuffer, af.GetResult(aggCtxs[i]))
+		}
+		e.mutableRow.SetDatums(e.rowBuffer...)
+		if chk.NumCols() == 0 {
```

Comment:
https://github.com/pingcap/tidb/pull/5563 has refactored `AppendRow` to handle virtual rows, we can just `chk.AppendRow(e.mutableRow.ToRow())`

Reply:
ok